File: Ruby/2024/2024-001/2024-001-part-002.rb
# Function to calculate the total similarity score
def calculate_similarity_score(file_path)
  # Read the file and load the location IDs into two arrays
  left_list = []
  right_list = []

  # Open the file and process each line
  File.foreach(file_path) do |line|
    # Split the line by spaces and convert the values to integers
    left, right = line.split.map(&:to_i)
    left_list << left
    right_list << right
  end

  # Create a hash to count occurrences of each number in the right list
  right_list_counts = Hash.new(0)
  right_list.each { |num| right_list_counts[num] += 1 }

  # Calculate the similarity score
  similarity_score = 0
  left_list.each do |left|
    similarity_score += left * right_list_counts[left]
  end

  # Output the similarity score
  puts "Total Similarity Score: #{similarity_score}"
end

# File path for input data
file_path = '\\\\vmware-host\\Shared Folders\\C\\advent-of-code-002\\Ruby\\2024\\2024-001\\input.txt'

# Call the function to calculate the similarity score
calculate_similarity_score(file_path)

File: other-programming-languages/Node-js/2024/2024-001/2024-001-part-002.js
const fs = require('fs');

// Function to calculate the similarity score
function calculateSimilarityScore(filePath) {
    // Read the file and split the content into lines
    const data = fs.readFileSync(filePath, 'utf-8').split('\n');

    // Initialize arrays for left and right lists
    let leftList = [];
    let rightList = [];

    // Process each line from the file
    data.forEach(line => {
        if (line.trim()) {
            const [left, right] = line.split(/\s+/).map(Number); // Split and convert to numbers
            leftList.push(left);
            rightList.push(right);
        }
    });

    // Create a dictionary to count occurrences of each number in the right list
    const rightListCounts = rightList.reduce((counts, num) => {
        counts[num] = (counts[num] || 0) + 1;
        return counts;
    }, {});

    // Calculate the similarity score
    const similarityScore = leftList.reduce((total, left) => {
        return total + (left * (rightListCounts[left] || 0));
    }, 0);

    // Output the similarity score
    console.log('Total Similarity Score:', similarityScore);
}

// File path for input data
const filePath = '\\\\vmware-host\\Shared Folders\\C\\advent-of-code-002\\node-js\\2024\\2024_001\\input.txt';

// Call the function to calculate the similarity score
calculateSimilarityScore(filePath);
